lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1015 @@
+    // include it, and it guards against mistakenly reusing cached modules
+    // across architectures.
+    H = hash_combine(H, SubInvocation.getLangOptions().Target.getArchName());
+
+    // The SDK path is going to affect how this module is imported, so include
+    // it.
+    H = hash_combine(H, SubInvocation.getSDKPath());
+
+    return llvm::APInt(64, H).toString(36, /*Signed=*/false);
+  }
+
+  /// Calculate an output filename in \p SubInvocation's cache path that
+  /// includes a hash of relevant key data.
+  void computeCachedOutputPath(const CompilerInvocation &SubInvocation,
+                               llvm::SmallString<256> &OutPath) {
+    OutPath = SubInvocation.getClangModuleCachePath();
+    llvm::sys::path::append(OutPath, SubInvocation.getModuleName());
+    OutPath.append("-");
+    OutPath.append(getCacheHash(SubInvocation));
+    OutPath.append(".");
+    auto OutExt = file_types::getExtension(file_types::TY_SwiftModuleFile);
+    OutPath.append(OutExt);
+  }
+
+  // Checks that a dependency read from the cached module is up to date compared
+  // to the interface file it represents. If it's up-to-date, return its status,
+  // so we can switch hash-based dependencies to mtimes after we've validated
+  // the hashes.
+  Optional<llvm::vfs::Status> dependencyIsUpToDate(const FileDependency &dep) {
+    auto Status = getStatusOfDependency(fs, dep.getPath(), interfacePath,
+                                        diags, diagnosticLoc);
+    if (!Status) return None;
+
+    // If the sizes differ, then we know the file has changed.
+    if (Status->getSize() != dep.getSize()) return None;
+
+    // Otherwise, if this dependency is verified by modification time, check
+    // it vs. the modification time of the file.
+    if (dep.isModificationTimeBased()) {
+      uint64_t mtime =
+        Status->getLastModificationTime().time_since_epoch().count();
+      return mtime == dep.getModificationTime() ? Status : None;
+    }
+
+    // Slow path: if the dependency is verified by content hash, check it vs.
+    // the hash of the file.
+    auto buf = getBufferOfDependency(fs, dep.getPath(), interfacePath,
+                                     diags, diagnosticLoc);
+    if (!buf) return None;
+
+    if (xxHash64(buf->getBuffer()) == dep.getContentHash())
+      return Status;
+
+    return None;
+  }
+
+  // Check if all the provided file dependencies are up-to-date compared to
+  // what's currently on disk.
+  bool dependenciesAreUpToDate(ArrayRef<FileDependency> deps) {
+    for (auto &in : deps) {
+      if (dependencyTracker)
+        dependencyTracker->addDependency(in.getPath(), /*isSystem*/false);
+      auto status = dependencyIsUpToDate(in);
+      if (!status) {
+        LLVM_DEBUG(llvm::dbgs() << "Dep " << in.getPath()
+                   << " is directly out of date\n");
+        return false;
+      }
+      LLVM_DEBUG(llvm::dbgs() << "Dep " << in.getPath() << " is up to date\n");
```

Comment:
That was leftover from a previous optimization. I threaded that status through all the way to the `forwardingModuleIsUpToDate` check, and then used it, so we only did one `stat`. It seemed like overkill, especially since I'd like to cache stats/hashes externally.